src/lib/crypto/kimchi_backend/common/plonk_constraint_system.ml

```diff
@@ -1331,47 +1528,253 @@ end = struct
+           ; Some (reduce_to_v v0p3)
+           ; Some (reduce_to_v v0p4)
+           ; Some (reduce_to_v v0p5)
+           ; Some (reduce_to_v v0c0)
+           ; Some (reduce_to_v v0c1)
+           ; Some (reduce_to_v v0c2)
+           ; Some (reduce_to_v v0c3)
+           ; Some (reduce_to_v v0c4)
+           ; Some (reduce_to_v v0c5)
+           ; Some (reduce_to_v v0c6)
+           ; Some (reduce_to_v v0c7) (* LSBs *)
           |]
         in
         let coeff = if Fp.equal compact Fp.one then Fp.one else Fp.zero in
         add_row sys vars RangeCheck0 [| coeff |]
+    | Plonk_constraint.T
+        (RangeCheck1
+          { (* Current row *) v2
+          ; v12
+          ; v2c0
+          ; v2p0
+          ; v2p1
+          ; v2p2
+          ; v2p3
+          ; v2c1
+          ; v2c2
+          ; v2c3
+          ; v2c4
+          ; v2c5
+          ; v2c6
+          ; v2c7
+          ; v2c8
+          ; (* Next row *) v2c9
+          ; v2c10
+          ; v2c11
+          ; v0p0
+          ; v0p1
+          ; v1p0
+          ; v1p1
+          ; v2c12
+          ; v2c13
+          ; v2c14
+          ; v2c15
+          ; v2c16
+          ; v2c17
+          ; v2c18
+          ; v2c19
+          } ) ->
+        (*
+        //!       0      1      2     3    4    5    6     7     8     9    10    11    12   13     14
+        //! Curr: v2   v12   v2c0  v2p0 v2p1 v2p2 v2p3  v2c1  v2c2  v2c3  v2c4  v2c5  v2c6 v2c7   v2c8
+        //! Next: v2c9 v2c10 v2c11 v0p0 v0p1 v1p0 v1p1 v2c12 v2c13 v2c14 v2c15 v2c16 v2c17 v2c18 v2c19
+        *)
+        let vars_curr =
+          [| (* Current row *) Some (reduce_to_v v2)
+           ; Some (reduce_to_v v12)
+           ; Some (reduce_to_v v2c0) (* MSBs *)
+           ; Some (reduce_to_v v2p0)
+           ; Some (reduce_to_v v2p1)
+           ; Some (reduce_to_v v2p2)
+           ; Some (reduce_to_v v2p3)
+           ; Some (reduce_to_v v2c1)
+           ; Some (reduce_to_v v2c2)
+           ; Some (reduce_to_v v2c3)
+           ; Some (reduce_to_v v2c4)
+           ; Some (reduce_to_v v2c5)
+           ; Some (reduce_to_v v2c6)
+           ; Some (reduce_to_v v2c7)
+           ; Some (reduce_to_v v2c8) (* LSBs *)
+          |]
+        in
+        let vars_next =
+          [| (* Next row *) Some (reduce_to_v v2c9)
+           ; Some (reduce_to_v v2c10)
+           ; Some (reduce_to_v v2c11)
+           ; Some (reduce_to_v v0p0)
+           ; Some (reduce_to_v v0p1)
+           ; Some (reduce_to_v v1p0)
+           ; Some (reduce_to_v v1p1)
+           ; Some (reduce_to_v v2c12)
+           ; Some (reduce_to_v v2c13)
+           ; Some (reduce_to_v v2c14)
+           ; Some (reduce_to_v v2c15)
+           ; Some (reduce_to_v v2c16)
+           ; Some (reduce_to_v v2c17)
+           ; Some (reduce_to_v v2c18)
+           ; Some (reduce_to_v v2c19)
+          |]
+        in
+        add_row sys vars_curr RangeCheck1 [||] ;
+        add_row sys vars_next Zero [||]
+    | Plonk_constraint.T
```

Comment:
The problem about this approach is that you don't give support for the final Zero gate for the bound check.

Reply:
Let's do this as gadget as you're suggesting

Reply:
Checked with Matthew and the circuit or gadget developer can modify it from zero subsequently.